### PR TITLE
Added more mips instructions and fixed immediate shift errors

### PR DESCRIPTION
**Pull request by SiD3W4y**
I didn't rebuild the bindings because there are errors in the csharp generator.

### CODE REVIEW

File: src/arch/mips.py [hunk 1]
```diff
@@ -65,10 +88,10 @@ def translate(self, input: IO[str]) -> Functions:
 
                     func = Function(name, [ param('address', TYPE_U32) ])
 
-                    code = Literal(opcode << 26, TYPE_U32)
-                    truncate_lit = Literal(0x3ffffff, TYPE_U32)
+                    opcode = int(chunks[2], 16)
 
-                    truncated = Binary(OP_BITWISE_AND, truncate_lit, Binary(OP_SHL, Var('address'), 2))
+                    code = Literal(opcode << 26, TYPE_U32)
+                    truncated = cast(Binary(OP_SHR, Var('address'), 2), 26)
```
Comment (71) [inline src/arch/mips.py]: You passed a number instead of an `Expression` here

Comment (71) [reply src/arch/mips.py]: ... And further above

File: src/arch/mips.py [hunk 2]
```diff
@@ -47,16 +50,36 @@ def translate(self, input: IO[str]) -> Functions:
...
                     func += Set(TYPE_U32, reduce(lambda a, b: Binary(OP_BITWISE_OR, a, b), vals))
 
                     yield func
                 
+                elif mode == 'RI':
+                    # type RI
+                    # mode for branches
+                    # (opcode: 6b) (register source: 5b) (funct: 5b) (imm: 16b)
+                    func = Function(name, [ param('rs', TYPE_MIPS_REG, TYPE_U32),
+                                            param('target', TYPE_U16, TYPE_U32)])
```
Comment (71) [inline src/arch/mips.py:69]: I'mma add you to add a single space between the first `)` and the following `]` at the end of the line, for clarity

File: src/arch/mips.py [hunk 3]
```diff
@@ -47,16 +50,36 @@ def translate(self, input: IO[str]) -> Functions:
...
+                    
+                    opcode = int(chunks[2], 16)
+                    fcnt = int(chunks[3], 16)
+                    
+                    vals = [
+                        Literal(opcode << 26, TYPE_U32),
+                        Binary(OP_SHL, cast(Var('rs'), 5), 16),
```
Comment (71) [inline src/arch/mips.py]: Number given instead of literal expression.

File: src/arch/mips.py [hunk 4]
```diff
@@ -47,16 +50,36 @@ def translate(self, input: IO[str]) -> Functions:
...
+                        cast(Binary(OP_SHR, Var('target'), 2), 16)
```
Comment (71) [inline src/arch/mips.py]: Number given instead of literal expression (again).

File: src/arch/mips.py [hunk 5]
```diff
@@ -85,12 +108,16 @@ def translate(self, input: IO[str]) -> Functions:
                                             param('imm', TYPE_U16, TYPE_U32)])
 
                     opcode = int(chunks[2], 16)
+                    immediate = cast(Var('imm'), 16)
+
+                    if name in ['beq', 'bne', 'blez', 'bgtz']:
+                        immediate = Binary(OP_SHR, immediate, 2)
```
Comment (71) [inline src/arch/mips.py:115]: Same here.

File: src/arch/mips.py [hunk 6]
```diff
@@ -24,6 +24,9 @@ def declarations(self) -> Declarations:
         yield DistinctType(TYPE_MIPS_REG, 'A Mips register.', [ Constant(n, i) for i, n in enumerate(mips_registers) ])
 
     def translate(self, input: IO[str]) -> Functions:
+        def cast(var, bits):
```
Comment (71) [inline src/arch/mips.py]: Any chance we'd get a small description of the function as doc comments?

**Review by 71 — CHANGES_REQUESTED** [review]
Thanks a lot, but there are a few small issues.

**Comment by SiD3W4y** [comment]
Should be fixed

**Review by 71 — APPROVED** [review]

**Pull request merged**